src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -551,6 +551,97 @@ public SolutionCompilationState WithProjectOutputFilePath(ProjectId projectId, s
+        }
+
+        return
+             WithProjectName(projectId, attributes.Name)
+            .WithProjectAssemblyName(projectId, attributes.AssemblyName)
+            .WithProjectFilePath(projectId, attributes.FilePath)
+            .WithProjectOutputFilePath(projectId, attributes.OutputFilePath)
+            .WithProjectOutputRefFilePath(projectId, attributes.OutputRefFilePath)
+            .WithProjectCompilationOutputInfo(projectId, attributes.CompilationOutputInfo)
+            .WithProjectDefaultNamespace(projectId, attributes.DefaultNamespace)
+            .WithHasAllInformation(projectId, attributes.HasAllInformation)
+            .WithRunAnalyzers(projectId, attributes.RunAnalyzers)
+            .WithProjectChecksumAlgorithm(projectId, attributes.ChecksumAlgorithm);
+    }
+
+    public SolutionCompilationState WithProjectInfo(ProjectInfo info)
+    {
+        var projectId = info.Id;
+        var newState = WithProjectAttributes(info.Attributes)
+            .WithProjectCompilationOptions(projectId, info.CompilationOptions)
+            .WithProjectParseOptions(projectId, info.ParseOptions)
+            .WithProjectReferences(projectId, info.ProjectReferences)
+            .WithProjectMetadataReferences(projectId, info.MetadataReferences)
+            .WithProjectAnalyzerReferences(projectId, info.AnalyzerReferences);
+
+        var oldProjectState = SolutionState.GetRequiredProjectState(projectId);
+
+        using var _1 = ArrayBuilder<DocumentInfo>.GetInstance(out var addedDocumentInfos);
+        using var _2 = ArrayBuilder<DocumentId>.GetInstance(out var removedDocumentInfos);
+
+        UpdateDocuments<DocumentState>(info.Documents);
+        UpdateDocuments<AdditionalDocumentState>(info.AdditionalDocuments);
+        UpdateDocuments<AnalyzerConfigDocumentState>(info.AnalyzerConfigDocuments);
+
+        void UpdateDocuments<TDocumentState>(IReadOnlyList<DocumentInfo> newDocumentInfos)
+            where TDocumentState : TextDocumentState
+        {
+            using var _3 = ArrayBuilder<TDocumentState>.GetInstance(out var updatedDocuments);
+
+            var oldDocumentStates = oldProjectState.GetDocumentStates<TDocumentState>();
+
+            foreach (var newDocumentInfo in newDocumentInfos)
+            {
+                if (oldDocumentStates.TryGetState(newDocumentInfo.Id, out var oldDocumentState))
+                {
+                    var newDocumentState = (TDocumentState)oldDocumentState.WithDocumentInfo(newDocumentInfo);
+                    if (oldDocumentState != newDocumentState)
+                    {
+                        updatedDocuments.Add(newDocumentState);
+                    }
+                }
+                else
+                {
+                    addedDocumentInfos.Add(newDocumentInfo);
+                }
+            }
+
+            if (!oldDocumentStates.Ids.IsEmpty())
+            {
+                var newDocumentsLookup = newDocumentInfos.ToLookup(static d => d.Id);
```

Comment:
dont' really need this as a lookup, do you?  it's more just a set.  could be .select(d => d.id).toset().